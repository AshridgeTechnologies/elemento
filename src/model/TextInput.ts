import BaseElement from './BaseElement'
import Element from './Element'
import {ElementId, PropertyValue} from './Types'

export type Properties = {
    readonly initialValue?: PropertyValue,
    readonly maxLength?: PropertyValue,
    readonly multiline?: PropertyValue,
    readonly label?: PropertyValue
}
export default class TextInput extends BaseElement<Properties> implements Element {

    constructor(
        id: ElementId,
        name: string,
        properties: Properties) {
        super(id, name, properties)
    }

    static is(element: Element): element is TextInput {
        return element.constructor.name === this.name
    }

    static defaultValue = ''

    get initialValue() { return this.properties.initialValue }
    get maxLength() { return this.properties.maxLength }
    get multiline() { return this.properties.multiline }
    get label() { return this.properties.label }
}