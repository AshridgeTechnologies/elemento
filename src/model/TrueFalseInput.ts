import Element from './Element'
import {PropertyType} from './Types'
import BaseInputElement, {BaseInputProperties} from './BaseInputElement'

export type Properties = BaseInputProperties<boolean>
export default class TrueFalseInput extends BaseInputElement<Properties> implements Element {

    static kind = 'TrueFalseInput'
    static get iconClass() { return 'check_box_outlined' }
    get valueType(): PropertyType { return 'boolean' }
}
