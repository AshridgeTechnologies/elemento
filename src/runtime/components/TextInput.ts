import React, {ChangeEvent} from 'react'
import {TextField} from '@mui/material'
import {definedPropertiesOf} from '../../util/helpers'
import {updateState} from '../appData'
import {valueOfProps} from '../runtimeFunctions'

type Properties = {state: {value?: string, _path: string, _controlValue: string | null}, label?: string, maxLength?: number, width?: string | number, multiline?: boolean}

export default function TextInput({state, ...props}: Properties) {
    const {maxLength, label, multiline, width} = valueOfProps(props)
    const inputProps = maxLength !== undefined ? {inputProps: {maxLength}} : {}
    const widthProp = width !== undefined ? {width} : {}
    const sxProps = {sx: {...widthProp}}
    const optionalProps = definedPropertiesOf({label, multiline})
    const {_path: path} = state
    const value = state._controlValue ?? ''
    const onChange = (event: ChangeEvent) => {
        const controlValue = (event.target as any).value
        const updateValue = controlValue !== '' ? controlValue : null
        updateState(path, {value: updateValue })
    }

    return React.createElement(TextField, {
        id: path,
        type: 'text',
        variant: 'outlined',
        size: 'small',
        value,
        onChange,
        ...inputProps,
        ...sxProps,
        ...optionalProps
    })
}
