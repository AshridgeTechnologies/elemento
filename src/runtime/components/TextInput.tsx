import React, {ChangeEvent, FocusEvent} from 'react'
import {TextField} from '@mui/material'
import {definedPropertiesOf} from '../../util/helpers'
import {PropVal, valueOfProps} from '../runtimeFunctions'
import InputComponentState from './InputComponentState'
import {useGetObjectState} from '../appData'
import {TextType} from '../../shared/types'
import {pick} from 'ramda'
import {InputWithInfo} from './InputWithInfo'


type Properties = {path: string, label?: PropVal<string>, width?: PropVal<string | number>, multiline?: PropVal<boolean>, readOnly?: PropVal<boolean>}

export default function TextInput({path, ...props}: Properties) {
    const {label, multiline: multilineProp, width, readOnly} = valueOfProps(props)
    const widthProp = width !== undefined ? {width} : {}
    const sxProps = {sx: {...widthProp}}

    const state = useGetObjectState<TextInputState>(path)
    const {value} = state
    const dataType = state.dataType
    const required = dataType?.required
    const multiline = dataType?.format === 'multiline' || multilineProp
    const multilineProps = multiline ? {minRows: 2, maxRows: 10} : {}
    const optionalProps = definedPropertiesOf({label, multiline, ...multilineProps})

    const dataTypeProps = definedPropertiesOf(pick(['maxLength', 'minLength'], dataType ?? {}))
    const format = dataType?.format
    if (format && ['url', 'email'].includes(format)) {
        dataTypeProps.type = format
    }
    const inputPropsValues = {...dataTypeProps, ...definedPropertiesOf({readOnly}), }
    const inputProps = Object.keys(inputPropsValues).length > 0 ? {inputProps: inputPropsValues} : {}

    const error = state.errorsShown && !state.valid
    const helperText = state.errorsShown && state.errors ? (state.errors as string[]).join('.  ') : undefined

    const onChange = (event: ChangeEvent) => {
        const controlValue = (event.target as any).value
        const updateValue = controlValue !== '' ? controlValue : null
        state._setValue(updateValue)
    }
    const onBlur = (_event: FocusEvent) => state.ShowErrors(true)

    const formControl =  React.createElement(TextField, {
        id: path,
        type: 'text',
        variant: 'outlined',
        size: 'small',
        value,
        error,
        helperText,
        onChange,
        onBlur,
        ...inputProps,
        ...sxProps,
        ...optionalProps
    })
    return InputWithInfo({description: dataType?.description, required, width, formControl})
}

export class TextInputState extends InputComponentState<string, TextType> {
    defaultValue = ''
}

TextInput.State = TextInputState