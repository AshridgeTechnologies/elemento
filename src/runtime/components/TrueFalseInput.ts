import {ChangeEvent, createElement} from 'react'
import {FormControlLabel, Switch} from '@mui/material'
import {valueOfProps} from '../runtimeFunctions'
import {proxyUpdateFnType} from '../stateProxy'
import {InputComponentState} from './InputComponentState'

type Properties = {state: {value?: boolean, _path: string, _controlValue: boolean | null, _update: proxyUpdateFnType}, label?: string}

export default function TrueFalseInput({state, ...props}: Properties) {
    const {_path: path} = state
    const value = state._controlValue ?? false
    const {label = ''} = valueOfProps(props)
    const onChange = (event: ChangeEvent) => state._update({value: (event.target as any).checked})

    return createElement(FormControlLabel, {
        label,
        labelPlacement: 'start',
        control: createElement(Switch, {
            id: path,
            size: 'small',
            color: 'primary',
            checked: value,
            onChange: onChange
        })
    })
}


TrueFalseInput.State = class State extends InputComponentState<boolean> {
    defaultValue = false
}