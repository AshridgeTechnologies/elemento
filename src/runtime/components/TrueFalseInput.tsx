import React, {ChangeEvent, createElement, FocusEvent} from 'react'
import {Checkbox, FormControl, FormControlLabel, FormHelperText} from '@mui/material'
import {valueOfProps} from '../runtimeFunctions'
import InputComponentState from './InputComponentState'
import {TrueFalseType} from '../types'
import {
    BaseInputComponentProperties,
    getLabelWithRequired,
    inputElementProps,
    propsForInputComponent,
    sxPropsForFormControl
} from './ComponentHelpers'
import {noop} from 'lodash'
import {useObject} from '../appStateHooks'
import {omit} from 'ramda'

type Properties = BaseInputComponentProperties

export default function TrueFalseInput({path, ...props}: Properties) {
    const {label = '', readOnly, show, styles = {}} = valueOfProps(props)
    const sx = sxPropsForFormControl(styles, show)

    const state = useObject<TrueFalseInputState>(path)
    const {dataValue, dataType} = state
    const value = dataValue ?? false
    const labelWithRequired = getLabelWithRequired(dataType, label)
    const inputComponentProps = propsForInputComponent(dataType, styles)
    const inputComponentPropsForCheckbox = omit(['endAdornment'], inputComponentProps.InputProps)
    const inputProps = inputElementProps(styles, false, {})

    const error = state.errorsShown && !state.valid
    const helperText = state.errorsShown && state.errors ? (state.errors as string[]).join('.  ') : undefined

    const onChange = (event: ChangeEvent) => {
        const controlValue = (event.target as any).checked
        const updateValue = controlValue !== '' ? controlValue : null
        state._setValue(updateValue)
    }
    const onBlur = (_event: FocusEvent) => state._setBlurred()

    const checkbox = createElement(Checkbox, {
        id: path,
        size: 'small',
        color: 'primary',
        checked: value,
        onChange: readOnly ? noop : onChange,
        disabled: readOnly,
        onBlur,
        ...inputProps,
        ...inputComponentPropsForCheckbox,
    })
    return <FormControl error={error} variant="standard">
        <FormControlLabel label={labelWithRequired}
                          labelPlacement='start'
                          htmlFor={path}
                          control={checkbox}
                          sx={sx}/>
        <FormHelperText>{helperText}</FormHelperText>
    </FormControl>
}

export class TrueFalseInputState extends InputComponentState<boolean, TrueFalseType> {
    defaultValue = false

    get modified() {
        const stateValue = this.state.value
        return stateValue !== undefined
            && (stateValue ?? false) !== (this.originalValue ?? false)
    }
}

TrueFalseInput.State = TrueFalseInputState
