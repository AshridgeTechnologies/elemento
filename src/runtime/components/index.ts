export {default as App} from './App'
export {default as Button} from './Button'
export {default as Icon} from './Icon'
export {default as Image} from './Image'
export {default as UserLogon} from './UserLogon'
export {default as Menu} from './Menu'
export {default as MenuItem} from './MenuItem'
export {default as NumberInput} from './NumberInput'
export {default as DateInput} from './DateInput'
export {default as TrueFalseInput} from './TrueFalseInput'
export {default as SpeechInput} from './SpeechInput'
export {default as Data} from './Data'
export {default as Calculation} from './Calculation'
export {default as Timer} from './Timer'
export {default as Collection} from './Collection'
export {default as Page} from './Page'
export {default as ComponentElement} from './ComponentElement'
export {default as Form} from './Form'
export {default as BaseFormState} from './FormState'
export {default as SelectInput} from './SelectInput'
export {default as TextElement} from './TextElement'
export {default as Layout} from './Layout'
export {default as Block} from './Block'
export {default as AppBar} from './AppBar'
export {default as ListElement} from './ListElement'
export {default as ListItem} from './ListItem'
export {default as ItemSet} from './ItemSet'
export {default as ItemSetItem} from './ItemSetItem'
export {default as TextInput} from './TextInput'
export {default as MemoryDataStore} from './MemoryDataStore'
export {default as FileDataStore} from './FileDataStore'
export {default as FileDataStoreImpl} from './FileDataStoreImpl'
export {default as IdbDataStoreImpl} from './IdbDataStoreImpl'
export {default as BrowserDataStore} from './BrowserDataStore'
export {default as FirestoreDataStore} from './FirestoreDataStore'
export {default as FirestoreDataStoreImpl} from './FirestoreDataStoreImpl'
export {default as ServerAppConnector} from './ServerAppConnector'
export {default as Adapter} from './Adapter'
