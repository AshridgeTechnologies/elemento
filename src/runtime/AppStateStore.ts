import SubscribableStore, {Callback, Id, UnsubscribeFn} from './SubscribableStore'
import {ComponentState} from './components/ComponentState'
import {unique} from 'radash'

export type StoredState = ComponentState<any>

export default class AppStateStore {

    constructor(private store: SubscribableStore = new SubscribableStore()) {
        this.subscribeToAllChanges()
    }

    get(id: string): StoredState {
        return this.store.get(id)
    }

    set<T extends StoredState>(id: string, item: T) {
        item.init(this.appStoreInterface(id), id)
        this.store.set(id, item)
    }

    updateIfChanged<T extends StoredState>(id: string, item: T) {
        const existingItem = this.get(id)
        const latestItem = existingItem ? existingItem.updateFrom(item) : item
        if (latestItem !== existingItem) {
            this.setDeferNotifications(id, latestItem)
        }
    }

    subscribe(id: string, callback: Callback): UnsubscribeFn {
        return this.store.subscribe(id, callback)
    }

    private subscribeToAllChanges() {
        this.store.subscribeAll( (ids: string[]) => this.updateParentItems(ids))
    }

    private updateParentItems(ids: Id[]) {
        const extractParent = (id: Id) => id.replace(/\.\w+$/, '')
        const parentIds = unique(ids.map(extractParent))
        parentIds.forEach( id => this.get(id)?.onChildStateChange() )
    }

    private setDeferNotifications<T extends StoredState>(id: string, item: T) {
        if (!this.store.isNotifyDeferred) {
            this.store.deferNotifications()
            setTimeout(() => this.store.sendNotifications(), 0)
        }
        this.set(id, item)
    }

    private appStoreInterface(id: string) {
        const store = this
        return {
            latest() {
                return store.get(id)
            },
            updateVersion(changes: object) {
                const updatedItem = store.get(id).withMergedState(changes)
                store.set(id, updatedItem)
            },
            getChildState(subPath: string) {
                return store.get(id + '.' + subPath)
            },
            getOrCreateChildState(subPath: string, item: StoredState) {
                store.updateIfChanged(id + '.' + subPath, item)
                return this.getChildState(subPath)
            },
            getApp() {
                const appId = id.split('.')[0]
                return store.get(appId)
            }
        }
    }
}
