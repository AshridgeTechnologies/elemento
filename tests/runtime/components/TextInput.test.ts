import {expect, MockedFunction, test, vi} from "vitest"
/**
 * @vitest-environment jsdom
 */
import {TextInput} from '../../../src/runtime/components/index'
import {componentJSON, snapshot, valueObj, wait, wrappedTestElementNew} from '../../testutil/testHelpers'
import {act, render} from '@testing-library/react'
import {actWait, testContainer} from '../../testutil/rtlHelpers'
import {TextInputState} from '../../../src/runtime/components/TextInput'
import {TrueFalseInputState} from '../../../src/runtime/components/TrueFalseInput'
import {TextType} from '../../../src/runtime/types'
import AppStateStore from '../../../src/runtime/state/AppStateStore'

const [textInput, appStoreHook] = wrappedTestElementNew(TextInput)

const stateAt = (path: string) => appStoreHook.stateAt(path)

const DEBUG_TIMEOUT = 1000000

test('TextInput element produces output with properties supplied', () => {
        const component = textInput('app.page1.width', {
            initialValue: 'Hi there!',
            maxLength: 11,
            width: 23,
            readOnly: true,
            label: 'Item Description',
            styles: {border: '1px solid green'}
        })
        expect(componentJSON(component)).toMatchSnapshot()
})

test('TextInput element produces output with properties supplied as state objects', () => {
    const {el} = testContainer(textInput('app.page1.widget1', {
        initialValue: 'Hello!',
        label: new TextInputState({initialValue: 'Item Description'}),
        readOnly: new TrueFalseInputState({value: true})
    }))
    expect(el`app.page1.widget1`.readOnly).toBe(true)
    expect(el`label[for="app.page1.widget1"]`.innerHTML).toBe('Item Description')
})

test('TextInput element produces output with multiline', () => {
    const {container} = render(textInput('app.page1.description', {initialValue: 'Hi there!', multiline: true, label: 'Item Description'}))
    expect(container.innerHTML).toMatchSnapshot()
})

test('TextInput element produces output with description info', () => {
    const textType = new TextType('tt1', {description: 'A text input for entering text'})
    const {container} = render(textInput('app.page1.description', {initialValue: 'Hi there!', dataType: textType, multiline: true, label: 'Item Description'}))
    expect(container.innerHTML).toMatchSnapshot()
})

test('keyAction function is called with key event', async () => {
    const keyAction = vi.fn()
    const {keyDown} = testContainer(textInput('app.page1.widget1', {initialValue: 'Hello!',
        label: 'Item Description',
        readOnly: false,
        keyAction
    }))
    await actWait( () => keyDown('app.page1.widget1', 'Enter'))
    expect(keyAction).toHaveBeenCalled()
    const callFirstArg: any = (keyAction as MockedFunction<any>).mock.calls[0][0]
    expect(callFirstArg.key).toBe('Enter')
})

test('TextInput element produces output with default values where properties omitted',
    snapshot(textInput('app.page1.height', {initialValue: undefined}))
)

test('TextInput shows value from the state supplied', () => {
    const {el} = testContainer(textInput('app.page1.widget1', {initialValue: 'Hello!'}))
    expect(el`app.page1.widget1`.value).toBe('Hello!')
})

test('TextInput shows empty value when state value is absent', () => {
    const {el} = testContainer(textInput('app.page1.widget1', {}))
    expect(el`app.page1.widget1`.value).toBe('')
})

test('TextInput shows empty value when state value is set to undefined', () => {
    const {el} = testContainer(textInput('app.page1.widget1', {initialValue: undefined}))
    expect(el`app.page1.widget1`.value).toBe('')
})

test('TextInput shows initial value when state value is set to undefined and initial value exists', () => {
    const {el} = testContainer(textInput('app.page1.widget1', {initialValue: 'Axe'}))
    expect(el`app.page1.widget1`.value).toBe('Axe')
})

test('TextInput shows initial value when state value is set to undefined and initial value is a value object', () => {
    const {el} = testContainer(textInput('app.page1.widget1', {initialValue: valueObj('Axe')}))
    expect(el`app.page1.widget1`.value).toBe('Axe')
})

test('TextInput shows empty value when state value is set to null and initial value exists', () => {
    const {el} = testContainer(textInput('app.page1.widget1', new TextInput.State({initialValue: 'Axe'})._withStateForTest({value: null})))
    expect(el`app.page1.widget1`.value).toBe('')
})

test('TextInput stores updated values in the app store section for its path', async () => {
    const {el, user} = testContainer(textInput('app.page1.sprocket', {initialValue: 'Hi'}))
    const inputEl = el`app.page1.sprocket`
    await user.type(inputEl, '!')
    await wait(10)
    expect(stateAt('app.page1.sprocket').value).toBe('Hi!')
})

test('TextInput stores null value in the app store when cleared', async () => {
    const {el, user} = testContainer(textInput('app.page1.sprocket2', {initialValue: 'Hi'}))

    const inputEl = el`app.page1.sprocket2`
    await user.clear(inputEl)
    await wait(10)
    expect(stateAt('app.page1.sprocket2').dataValue).toBe(null)
})

test('TextInput stores errors shown when blurred and ShowErrors or Reset called', async () => {
    const {el, user, domContainer} = testContainer(textInput('app.page1.sprocket2', {initialValue: 'Hi'}))

    const inputEl = el`app.page1.sprocket2`
    await user.click(inputEl)
    await wait(10)
    expect(stateAt('app.page1.sprocket2').errorsShown).toBe(false)

    await user.click(domContainer.ownerDocument.body)
    await wait(10)
    expect(stateAt('app.page1.sprocket2').errorsShown).toBe(true)

    await act( () => stateAt('app.page1.sprocket2').ShowErrors(false) )
    expect(stateAt('app.page1.sprocket2').errorsShown).toBe(false)

    await act( () => stateAt('app.page1.sprocket2').ShowErrors(true) )
    expect(stateAt('app.page1.sprocket2').errorsShown).toBe(true)

    await act( () => stateAt('app.page1.sprocket2').Reset() )
    expect(stateAt('app.page1.sprocket2').errorsShown).toBe(false)
})

test('TextInput shows errors if errorShown is true', async () => {
    const minLength3 = new TextType('tt1', {minLength: 3})
    const {el} = testContainer(textInput('app.page1.sprocket2', {initialValue: 'Hi', label: 'The Input', dataType: minLength3}))

    await actWait( () => stateAt('app.page1.sprocket2').ShowErrors(false) )
    expect(el`p.MuiFormHelperText-root.Mui-error`).toBe(null)
    expect(el`label`).not.toHaveClass('Mui-error')

    await actWait( () => stateAt('app.page1.sprocket2').ShowErrors(true) )
    // expect(el`p.MuiFormHelperText-root.Mui-error`.innerHTML).toBe('Minimum length 3')
    expect(el`label`).toHaveClass('Mui-error')
})

test('TextInput shows required', async () => {
    const requiredType = new TextType('tt1', {required: true})

    const {el} = testContainer(textInput('app.page1.sprocket2', {initialValue: 'Hi', label: 'The Input', dataType: requiredType}))
    expect(el`label`.textContent).toBe('The Input *')
})

test('TextInput uses properties from dataType', async () => {
    const textType = new TextType('tt1', {minLength: 2, maxLength: 10, format: 'url'})
    const {el} = testContainer(textInput('app.page1.sprocket2', {initialValue: 'Hi', label: 'The Input', dataType: textType}))

    expect(el`input`.minLength).toBe(2)
    expect(el`input`.maxLength).toBe(10)
    expect(el`input`.type).toBe('url')
})

test('TextInput uses multiline from dataType', async () => {
    const textType = new TextType('tt1', {format: 'multiline'})
    const {el} = testContainer(textInput('app.page1.sprocket2', {initialValue: 'Hi', label: 'The Input', dataType: textType}))
    expect(el`textarea`.textContent).toBe('Hi')
})

test('TextInput overrides properties from dataType', async () => {
    const {el} = testContainer(textInput('app.page1.sprocket2', {initialValue: 'Hi', label: 'The Input', multiline: true}))

    const textType = new TextType('tt1', {minLength: 2, maxLength: 10})
    await actWait(10)
    await actWait( () => stateAt('app.page1.sprocket2').props.dataType = textType )

    await actWait( () => stateAt('app.page1.sprocket2').Reset() )
    expect(el`textarea`.textContent).toBe('Hi')
})

test('TextInput can be focused', async () => {
    const elementId = 'app.page1.sprocket2'
    const {el} = testContainer(textInput(elementId, {initialValue: 'Hi', label: 'The Input'}))
    await actWait(10)
    expect(document.activeElement).not.toBe(el`app.page1.sprocket2`)
    await actWait( () => stateAt(elementId).Focus() )
    expect(document.activeElement).toBe(el`app.page1.sprocket2`)
    expect(stateAt(elementId).domElement).toBe(el`app.page1.sprocket2`)
})

test('State class has correct properties and functions with defaults', () => {
    const state = new TextInput.State({})
    expect(state.value).toBe('')
    expect(state.toString()).toBe('')
    expect(state.defaultValue).toBe('')
})

test('State class has correct properties and functions', async () => {
    const store = new AppStateStore()
    const state: TextInputState = store.getOrCreate('id1', TextInputState, {initialValue: 'car'})
    expect(state.value).toBe('car')
    expect(state.toString()).toBe('car')
    expect(state.defaultValue).toBe('')

    state.ShowErrors(true)
    await wait()
    expect(store.get('id1').errorsShown).toBe(true)
})

test('State class gives correct value when its value is a value object', () => {
    const state = new TextInput.State({initialValue: valueObj('car')})
    expect(state.value).toBe('car')
    expect(state.originalValue).toBe('car')
})

test('State class gives correct value when its value is another value whose value is a value object', () => {
    const state1 = new TextInput.State({initialValue: valueObj('car')})
    const state2 = new TextInput.State({initialValue: state1})
    expect(state2.value).toBe('car')
})

test('valueOf returns the value', () => {
    const state = new TextInput.State({initialValue: 'car'})
    expect(state.valueOf()).toBe('car')
})

test('string conversion uses the value', () => {
    const state = new TextInput.State({initialValue: 'car'})
    expect('x' + state).toBe('xcar')
})

test('Can use state object proxy in simple expressions', () => {
    const store = new AppStateStore()
    const state = store.getOrCreate('id1', TextInputState, {initialValue: 'car'})
    expect(state + 'park').toBe('carpark')
})

test('State is valid and has no errors when it has no data type', () => {
    const state = new TextInput.State({initialValue: 'car'})
    expect(state.valid).toBe(true)
    expect(state.errors).toBe(null)
})

test('State is valid when its value is valid for the data type', () => {
    const textType = new TextType('tt1', {minLength: 3})
    const stateValid = new TextInput.State({initialValue: 'car', dataType: textType})
    const stateInvalid = new TextInput.State({initialValue: 'ca', dataType: textType})
    expect(stateValid.dataType).toBe(textType)
    expect(stateValid.valid).toBe(true)
    expect(stateValid.errors).toBe(null)
    expect(stateInvalid.valid).toBe(false)
    expect(stateInvalid.errors).toStrictEqual(['Minimum length 3'])
})

test('State uses props or state value for validation and not default value', () => {
    const textType = new TextType('tt1', {required: true})
    const stateValid = new TextInput.State({initialValue: 'car', dataType: textType})
    const stateInvalidProps = new TextInput.State({initialValue: null, dataType: textType})
    const stateInvalidState = new TextInput.State({initialValue: 'car', dataType: textType})._withStateForTest({value: null})
    expect(stateValid.errors).toBe(null)
    expect(stateInvalidProps.errors).toStrictEqual(['Required'])
    expect(stateInvalidState.errors).toStrictEqual(['Required'])

})

test('State is modified only when its value is not null and different to its empty initial value', () => {
    const state = new TextInput.State({})
    expect(state.modified).toBe(false)
    const updatedState = state._withStateForTest({value: 'new car'})
    expect(updatedState.modified).toBe(true)
    const updatedAgainState = updatedState._withStateForTest({value: undefined})
    expect(updatedAgainState.modified).toBe(false)
    const updatedToNullState = updatedState._withStateForTest({value: null})
    expect(updatedToNullState.modified).toBe(false)
})

test('State is modified only when its value is not null and different to its null initial value', () => {
    const state = new TextInput.State({initialValue: null})
    expect(state.modified).toBe(false)
    const updatedState = state._withStateForTest({value: 'new car'})
    expect(updatedState.modified).toBe(true)
    const updatedAgainState = updatedState._withStateForTest({value: undefined})
    expect(updatedAgainState.modified).toBe(false)
    const updatedToNullState = updatedState._withStateForTest({value: null})
    expect(updatedToNullState.modified).toBe(false)
})

test('State is modified only when its value is different to its initial value', () => {
    const state = new TextInput.State({initialValue: 'car'})
    expect(state.modified).toBe(false)
    const updatedState = state._withStateForTest({value: 'new car'})
    expect(updatedState.modified).toBe(true)
    const updatedAgainState = updatedState._withStateForTest({value: 'car'})
    expect(updatedAgainState.modified).toBe(false)
    const updatedToNullState = updatedState._withStateForTest({value: null})
    expect(updatedToNullState.modified).toBe(true)
})

test('State stores errorsShown', () => {
    const state = new TextInput.State({initialValue: 'car'})
    expect(state.errorsShown).toBe(false)
    const updatedState = state._withStateForTest({errorsShown: true})
    expect(updatedState.errorsShown).toBe(true)
})

