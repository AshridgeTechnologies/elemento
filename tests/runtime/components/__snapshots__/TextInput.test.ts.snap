// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput element produces output with default values where properties omitted 1`] = `
<div
  className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
>
  <div
    className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
    onClick={[Function]}
  >
    <input
      aria-invalid={false}
      autoFocus={false}
      className="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
      disabled={false}
      id="app.page1.height"
      onAnimationStart={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      required={false}
      type="text"
      value=""
    />
    <fieldset
      aria-hidden={true}
      className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
    >
      <legend
        className="css-hdw1oc"
      >
        <span
          className="notranslate"
        >
          â€‹
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;

exports[`TextInput element produces output with multiline 1`] = `"<div class=\\"MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root\\"><label class=\\"MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1sumxir-MuiFormLabel-root-MuiInputLabel-root\\" data-shrink=\\"true\\" for=\\"app.page1.description\\" id=\\"app.page1.description-label\\">Item Description</label><div class=\\"MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-multiline css-1qxw4jt-MuiInputBase-root-MuiOutlinedInput-root\\"><textarea aria-invalid=\\"false\\" id=\\"app.page1.description\\" class=\\"MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputMultiline MuiInputBase-inputSizeSmall css-12tl3rr-MuiInputBase-input-MuiOutlinedInput-input\\" style=\\"height: 0px; overflow: hidden;\\">Hi there!</textarea><textarea aria-hidden=\\"true\\" class=\\"MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputMultiline MuiInputBase-inputSizeSmall css-12tl3rr-MuiInputBase-input-MuiOutlinedInput-input\\" readonly=\\"\\" tabindex=\\"-1\\" style=\\"visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;\\"></textarea><fieldset aria-hidden=\\"true\\" class=\\"MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline\\"><legend class=\\"css-1in441m\\"><span>Item Description</span></legend></fieldset></div></div>"`;

exports[`TextInput element produces output with properties supplied 1`] = `
<div
  className="MuiFormControl-root MuiTextField-root css-e7xw0c-MuiFormControl-root-MuiTextField-root"
>
  <label
    className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
    data-shrink={true}
    htmlFor="app.page1.width"
    id="app.page1.width-label"
  >
    Item Description
  </label>
  <div
    className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
    onClick={[Function]}
  >
    <input
      aria-invalid={false}
      autoFocus={false}
      className="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
      disabled={false}
      id="app.page1.width"
      maxLength={10}
      onAnimationStart={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      required={false}
      type="text"
      value="Hi there!"
    />
    <fieldset
      aria-hidden={true}
      className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
    >
      <legend
        className="css-1in441m"
      >
        <span>
          Item Description
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;
